### PR TITLE
[MXNET-347] Logical Operators AND, XOR, OR

### PR DESCRIPTION
**Pull request by anirudhacharya**
## Description ##
Adds Logical AND, XOR, OR operator.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] AND, OR, XOR operator.

## Comments ##
- @haojin2 @eric-haibin-lin @cjolivier01

### CODE REVIEW

File: src/operator/tensor/elemwise_binary_broadcast_op_logic.cc [hunk 1]
```diff
@@ -137,5 +137,59 @@ Example::
...
+Example::
+
+   x = [[ 1.,  1.,  1.],
+        [ 1.,  1.,  1.]]
+
+   y = [[ 0.],
+        [ 1.]]
+
+   broadcast_logical_and(x, y) = [[ 0.,  0.,  0.],
+                                   [ 1.,  1.,  1.]]
```
Comment (haojin2) [inline src/operator/tensor/elemwise_binary_broadcast_op_logic.cc]: Please align the lines

File: src/operator/tensor/elemwise_binary_broadcast_op_logic.cc [hunk 2]
```diff
@@ -137,5 +137,59 @@ Example::
...
+Example::
+
+   x = [[1  , 2,    0],
+        [0.5, -2.3, 0.0]]
+
+   y = [[2.0],
+        [0]]
+
+   broadcast_logical_or(x, y) = [[1. 1. 1.]
+ 								[1. 1. 0.]]
```
Comment (haojin2) [inline src/operator/tensor/elemwise_binary_broadcast_op_logic.cc]: Same here, please align the lines.

File: src/operator/tensor/elemwise_binary_broadcast_op_logic.cc [hunk 3]
```diff
@@ -137,5 +137,59 @@ Example::
...
+MXNET_OPERATOR_REGISTER_BINARY_BROADCAST(broadcast_logical_or)
+.describe(R"code(Returns the result of element-wise **logical or** with broadcasting.
+
+Example::
+
+   x = [[1  , 2,    0],
+        [0.5, -2.3, 0.0]]
+
+   y = [[2.0],
+        [0]]
```
Comment (haojin2) [inline src/operator/tensor/elemwise_binary_broadcast_op_logic.cc]: Why do the lines of input have different data types? Making all of them to be float is fine, please refer to examples of other existing ops.

File: src/operator/tensor/elemwise_binary_broadcast_op_logic.cc [hunk 4]
```diff
@@ -137,5 +137,59 @@ Example::
...
+
+   broadcast_logical_xor(x, y) = [[0. 0. 1.]
+ 								 [1. 1. 0.]]
```
Comment (haojin2) [inline src/operator/tensor/elemwise_binary_broadcast_op_logic.cc]: Same here, please align the lines.

File: src/operator/mshadow_op.h [hunk 5]
```diff
@@ -317,6 +317,12 @@ MXNET_BINARY_MATH_OP_NC(eq, a == b ? DType(1) : DType(0));
 
 MXNET_BINARY_MATH_OP_NC(ne, a != b ? DType(1) : DType(0));
 
+MXNET_BINARY_MATH_OP(logical_and, a && b ? DType(1) : DType(0));
+
+MXNET_BINARY_MATH_OP(logical_or, a || b ? DType(1) : DType(0));
+
+MXNET_BINARY_MATH_OP(logical_xor, (a || b) && !(a && b) ? DType(1) : DType(0));
```
Comment (cjolivier01) [inline src/operator/mshadow_op.h]: does anything use this?

File: tests/python/unittest/test_operator.py [hunk 6]
```diff
@@ -1581,6 +1581,27 @@ def test_bmin(a, b):
         data = gen_broadcast_data(idx=200)
         check_bmaxmin_gradient(c, data[0], data[1], 0.001, 1e-2, 1e-3)
 
+    def test_band(a, b):
+        c = mx.sym.broadcast_logical_and(a, b)
+        check_binary_op_forward(c, lambda x, y: np.logical_and(x, y), gen_broadcast_data, mx_nd_func=mx.nd.broadcast_logical_and)
```
Comment (cjolivier01) [inline tests/python/unittest/test_operator.py]: Why aren't there also non-broadcast versions (which will be much faster) and then you can just name it logical_or and use broadcast in the python when the shapes don't match similar to NDArray add?  Seems cumbersome that the numpy will be logical_or and for mxnet, you have to remember it's broadcast_logical_or

File: src/operator/mshadow_op.h (hunk 5 above)
Comment (anirudhacharya) [reply src/operator/mshadow_op.h]: meaning? use cases for having an implementation of these operators?

Comment (anirudhacharya) [reply src/operator/mshadow_op.h]: https://github.com/apache/incubator-mxnet/issues/9762

File: tests/python/unittest/test_operator.py (hunk 6 above)
Comment (haojin2) [reply tests/python/unittest/test_operator.py]: The broadcast_xxx function actually wraps the other function which is the elemwise_xxx. If we do want an interface for user to call logical_xxx without the broadcast, we can do it in Python just like how add/subtract/multiply/divide is implemented.

Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: Isn't add/sub etc.. also using broadcast_add, broadcast_sub ... http://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mathematical-functions ?

Comment (haojin2) [reply tests/python/unittest/test_operator.py]: @anirudhacharya Please take a look at: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/ndarray/ndarray.py#L2479

Comment (piiswrong) [reply tests/python/unittest/test_operator.py]: I think there should be a non broadcast version

Comment (anirudhacharya) [reply tests/python/unittest/test_operator.py]: Will add it.

**Comment by anirudhacharya** [comment]
@piiswrong @cjolivier01 could please take a look. I have addressed your comments.

**Comment by anirudhacharya** [comment]
@eric-haibin-lin should I update these under ``Comparison`` or should I create a new ``Logical`` category? I think even this - https://github.com/apache/incubator-mxnet/pull/10639 needs to be updated in the doc API.

**Pull request merged**